lib/SILOptimizer/Mandatory/TFPartition.cpp

```diff
@@ -2945,42 +2950,32 @@ bool PartitionCloner::finalizeOriginal() {
   for (unsigned idx = 0; idx != instructionsToRemove.size(); ++idx) {
     auto inst = instructionsToRemove[idx];
 
-    // When removing a builtin tensor op instruction TO, for each tensor
-    // argument TA of it, if TA is NOT generated by an instruction that we are
-    // to remove from host (e.g. other tensor ops or by TupleExtractInsts), emit
-    // a strong_release to balance the already emitted strong_retain's for
-    // TA. Recall the calling convention of TO is that TO consumes TA at +1.
-    //
-    // For example, before removing instruction
-    // %r1 = builtin "__tfop_Add"(%arg1, %arg2),
-    // we first emit strong_release'es on %arg1 and %arg2.
-    // If there is a subsequent instruction
-    // %r2 = builtin "__tfop_Add"(%r1, %arg3),
-    // we only emit strong_release on %arg3, and not on %r1, since the
-    // strong_retain on %r1 will be removed when all references of the first Add
-    // instruction are dropped along with that instruction itself.
-    //
-    // The same is true if the instruction defining %r1 is relaced with:
-    // %r1 = tuple_extract %arg1 : $(TensorHandle<?>, TensorHandle<?>), 0
-    if (isa<ApplyInst>(inst) || isa<BuiltinInst>(inst)) {
-      SILBuilder BH(inst);  // Builder for the host.
-      for (auto &operand : inst->getAllOperands()) {
-        auto op = operand.get();
+#ifndef NDEBUG
+    // When removing a builtin or ApplyInst tensor op instruction TO, for each tensor
+    // argument TA of it, since we know TA is taken at +0, we need not rebalance
+    // the retain/release count of TA.
+    if (isa<BuiltinInst>(inst)) {
+      // The tfop inst takes tensor operands as +0, but we know of no API that
+      // we can use to verify that here.
+    } else if (auto *ai = dyn_cast<ApplyInst>(inst)) {
+      // Sanity check: Verify that `inst` takes all tensor operands at +0.
+      auto *fri = cast<FunctionRefInst>(ai->getOperand(0));
```

Comment:
Is the operand `function_ref` in all possible cases? Assertions from `cast` are quite common hard to debug during model building. Just to confirm :)

Reply:
Done. :)